File: fsharp/rna-transcription/RNATranscriptionTest.fs
﻿module RNATranscriptionTests

open NUnit.Framework

open RNATranscription
    
[<Test>]
let ``Rna complement of cytosine is guanine`` () =
    Assert.That(toRna "C", Is.EqualTo("G"))

[<Test>]
let ``Rna complement of guanine is cytosine`` () =
    Assert.That(toRna "G", Is.EqualTo("C"))

[<Test>]
let ``Rna complement of thymine is adenine`` () =
    Assert.That(toRna "T", Is.EqualTo("A"))

[<Test>]
let ``Rna complement of adenine is uracil`` () =
    Assert.That(toRna "A", Is.EqualTo("U"))

[<Test>]
let ``Rna complement`` () =
    Assert.That(toRna "ACGTGGTCTTAA", Is.EqualTo("UGCACCAGAAUU"))

File: typescript/rna-transcription/rna-transcription.test.ts
import { toRna } from "./rna-transcription";

describe("Transcriptor", () => {
  it("transcribes cytosine to guanine", () => {
    expect(toRna("C")).toEqual("G");
  });

  it("transcribes guanine to cytosine", () => {
    expect(toRna("G")).toEqual("C");
  });

  it("transcribes adenine to uracil", () => {
    expect(toRna("A")).toEqual("U");
  });

  it("transcribes thymine to adenine", () => {
    expect(toRna("T")).toEqual("A");
  });

  it("transcribes all dna nucleotides to their rna complements", () => {
    expect(toRna("ACGTGGTCTTAA")).toEqual("UGCACCAGAAUU");
  });

  it("correctly handles invalid input", () => {
    expect(() => toRna("U")).toThrowError("Invalid input DNA.");
  });

  it("correctly handles completely invalid input", () => {
    expect(() => toRna("XXX")).toThrowError("Invalid input DNA.");
  });

  it("correctly handles partially invalid input", () => {
    expect(() => toRna("ACGTXXXCTTAA")).toThrowError("Invalid input DNA.");
  });
});
